clean_user_registry_secrets/clean_user_registry_secrets.py

```diff
@@ -0,0 +1,151 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Scripts used to remove user registry secrets in k8s"""
+
+from datetime import datetime, timedelta
+from pathlib import Path
+import argparse
+import re
+import logging
+from functools import reduce
+
+from kubernetes import client
+from kubernetes.config.incluster_config import (
+    SERVICE_CERT_FILENAME,
+    SERVICE_TOKEN_FILENAME,
+    InClusterConfigLoader,
+)
+
+
+def find_pod_by_secret(secret, k8s_client):
+    """Find the user jupyterhub podname based on the registry pull secret."""
+    label_keys = ['renku.io/commit-sha', 'renku.io/projectName', 'renku.io/username']
+    label_selector = []
+    for label_key in label_keys:
+        label_selector.append(f"{label_key}={secret.metadata.labels[label_key]}")
+    label_selector = ",".join(label_selector)
+
+    pod_list = k8s_client.list_namespaced_pod(
+        secret.metadata.namespace,
+        label_selector=label_selector,
+    )
+    if len(pod_list.items) > 0:
+        return pod_list.items[0].metadata.name
+    return None
+
+
+def remove_user_registry_secret(namespace, k8s_client, min_secret_age_hrs=0.25):
+    """Used in a cronjob to periodically remove old user registry secrets"""
+    secret_name_regex = ".+-registry-[a-z0-9-]{36}$"
+    label_keys = ['renku.io/commit-sha', 'renku.io/projectName', 'renku.io/username']
+    logging.info(
+        f"Checking for user registry secrets whose "
+        f"names match the regex: {secret_name_regex}"
+    )
+    secret_list = k8s_client.list_namespaced_secret(namespace)
+    for secret in secret_list.items:
+        # loop through secrets and find ones that match the predefined regex
+        secret_name = secret.metadata.name
+        secret_name_match = re.match(secret_name_regex, secret_name)
+        # calculate secret age
+        tz = secret.metadata.creation_timestamp.tzinfo
+        secret_age = datetime.now(tz=tz) - secret.metadata.creation_timestamp
+        min_secret_age = timedelta(hours=min_secret_age_hrs)
+        if (
+            secret_name_match is not None
+            and secret.type == "kubernetes.io/dockerconfigjson"
+            and secret.metadata.labels.get("component") == "singleuser-server"
+            and reduce(  # check that label keys for sha, project and username are present
+                lambda acc, val: acc and (val in secret.metadata.labels.keys()),
+                label_keys,
+                True,
+            )
+        ):
+            podname = find_pod_by_secret(secret, k8s_client)
+            if podname is None:
+                # pod does not exist, delete if secret is old enough
+                if secret_age > min_secret_age:
```

Comment:
if no pod using the secret exists, why do you care how old the secret is? To avoid the race condition where the culling process would catch a secret in between secret creation and pod creation? In that case the TTL of the secret can be much shorter imo

Reply:
@rokroskar yes I added this to prevent such a scenario. The secrets are fully unique now so there is no danger of rewriting something by mistake. So yeah my idea was better to have some secrets hang around for some time before they are deleted rather than delete them too quickly. 

As for you suggestion of using a TTL can you set a TTL on a secret? I don't see anything like that in documentation. Or maybe I misunderstand what you mean and you were suggesting to do something else?

Reply:
Oh I think I know what you mean - you are saying set the `-a` parameter (i.e. the max secret age) should be less than 15 minutes? I will set it to 5 minutes instead.

Reply:
yes that's what I meant - i don't see a scenario where it should me more than a few ms between the secret being created and the pod getting dispatched so even 5 minutes seems like a generous lifetime.

Reply:
@rokroskar I can go down to a lower value, 1 minute?

Reply:
haha no I think it's fine, I don't think it will cause any problems. If we see that the secrets are piling up too much for some reason we can reduce it further.